src/colmap/estimators/covariance.cc

```diff
@@ -0,0 +1,582 @@
+  int index1 = GetBlockIndex(params1);
+  int num_params_block1 = GetBlockTangentSize(params1);
+  int index2 = GetBlockIndex(params2);
+  int num_params_block2 = GetBlockTangentSize(params2);
+  return cov_poses_.block(index1, index2, num_params_block1, num_params_block2);
+}
+
+Eigen::MatrixXd BundleAdjustmentCovarianceEstimatorBase::GetCovariance(
+    double* params) const {
+  THROW_CHECK(HasValidFullCovariance());
+  THROW_CHECK(HasBlock(params));
+  int index = GetBlockIndex(params);
+  int num_params = GetBlockTangentSize(params);
+  return cov_variables_.block(index, index, num_params, num_params);
+}
+
+Eigen::MatrixXd BundleAdjustmentCovarianceEstimatorBase::GetCovariance(
+    const std::vector<double*>& blocks) const {
+  THROW_CHECK(HasValidFullCovariance());
+  std::vector<int> indices;
+  for (const double* block : blocks) {
+    THROW_CHECK(HasBlock(block));
+    int index = GetBlockIndex(block);
+    int num_params_pose = GetBlockTangentSize(block);
+    for (int i = 0; i < num_params_pose; ++i) {
+      indices.push_back(index + i);
+    }
+  }
+  size_t n_indices = indices.size();
+  Eigen::MatrixXd output(n_indices, n_indices);
+  for (size_t i = 0; i < n_indices; ++i) {
+    for (size_t j = 0; j < n_indices; ++j) {
+      output(i, j) = cov_variables_(indices[i], indices[j]);
+    }
+  }
+  return output;
+}
+
+Eigen::MatrixXd BundleAdjustmentCovarianceEstimatorBase::GetCovariance(
+    double* params1, double* params2) const {
+  THROW_CHECK(HasValidFullCovariance());
+  THROW_CHECK(HasBlock(params1));
+  THROW_CHECK(HasBlock(params2));
+  int index1 = GetBlockIndex(params1);
+  int num_params_block1 = GetBlockTangentSize(params1);
+  int index2 = GetBlockIndex(params2);
+  int num_params_block2 = GetBlockTangentSize(params2);
+  return cov_variables_.block(
+      index1, index2, num_params_block1, num_params_block2);
+}
+
+bool BundleAdjustmentCovarianceEstimatorCeresBackend::ComputeFull() {
+  ceres::Covariance::Options options;
+  ceres::Covariance covariance_computer(options);
+  std::vector<const double*> parameter_blocks;
+  parameter_blocks.insert(
+      parameter_blocks.end(), pose_blocks_.begin(), pose_blocks_.end());
+  parameter_blocks.insert(parameter_blocks.end(),
+                          other_variables_blocks_.begin(),
+                          other_variables_blocks_.end());
+  if (!covariance_computer.Compute(parameter_blocks, problem_)) return false;
+  int num_params = num_params_poses_ + num_params_other_variables_;
+  Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> covs(
+      num_params, num_params);
+  covariance_computer.GetCovarianceMatrixInTangentSpace(parameter_blocks,
+                                                        covs.data());
+  cov_variables_ = covs;
+  cov_poses_ = cov_variables_.block(0, 0, num_params_poses_, num_params_poses_);
+  return true;
+}
+
+bool BundleAdjustmentCovarianceEstimatorCeresBackend::Compute() {
```

Comment:
Would it make sense to call the two methods "ComputePoseCov" and "ComputePoseAndPointCov" ?

Reply:
The first method is to indeed compute pose covariance, but the second is to compute the covariance of all variables except for the points, as described in the header, and I felt that ComputePoseAndOtherVariablesCov is too long here.